src/ucp/proto/proto_am.inl

```diff
@@ -83,28 +83,96 @@ err:
+static UCS_F_ALWAYS_INLINE
+size_t ucp_dt_iov_copy_uct(uct_iov_t *iov, size_t *iovcnt, size_t max_dst_iov,
+                           ucp_frag_state_t *state, const ucp_dt_iov_t *src_iov,
+                           ucp_datatype_t datatype, size_t length_max)
+{
+    size_t iovcnt_offset, iov_offset, max_src_iov;
+    size_t src_it, src_it_offset, dst_it;
+    const uct_mem_h *memh;
+    size_t length_it = 0;
+
+    switch (datatype & UCP_DATATYPE_CLASS_MASK) {
+    case UCP_DATATYPE_CONTIG:
+        iov[0].buffer = (void *)src_iov + state->offset;
+        iov[0].length = length_max;
+        iov[0].memh   = state->dt.contig.memh;
+        iov[0].stride = 0;
+        iov[0].count  = 1;
+
+        *iovcnt   = 1;
+        length_it = iov[0].length;
+        break;
+    case UCP_DATATYPE_IOV:
+        memh          = state->dt.iov.memh;
+        iovcnt_offset = state->dt.iov.iovcnt_offset;
+        iov_offset    = state->dt.iov.iov_offset;
+        max_src_iov   = state->dt.iov.iovcnt;
+
+        for (src_it = 0, dst_it = 0; (dst_it < max_dst_iov) &&
+                                     ((src_it + iovcnt_offset) < max_src_iov) &&
+                                     !(length_it >= length_max); ++src_it) {
+            src_it_offset = src_it + iovcnt_offset;
+            if (src_iov[src_it_offset].length) {
+                iov[dst_it].buffer  = src_iov[src_it_offset].buffer + iov_offset;
+                iov[dst_it].length  = src_iov[src_it_offset].length - iov_offset;
+                iov[dst_it].memh    = memh[src_it_offset];
+                iov[dst_it].stride  = 0;
+                iov[dst_it].count   = 1;
+
+                length_it += iov[dst_it].length;
+                ++dst_it;
+            }
+            iov_offset = 0;
+        }
+
+        if (dst_it > 1) { /* the first src_iov buffer is not bigger than limit */
+            state->dt.iov.iov_offset     = iov_offset;
+        }
+        state->dt.iov.iovcnt_offset     += src_it;
+
+        if (length_it >= length_max) {
+            iov[dst_it - 1].length      -= (length_it - length_max);
+            length_it                    = length_max;
+            state->dt.iov.iov_offset    += iov[dst_it - 1].length;
+            state->dt.iov.iovcnt_offset -= 1;
+        }
+        *iovcnt = dst_it;
+
+        break;
+    default:
+        ucs_error("Invalid data type");
+    }
+
+    return length_it;
+}
+
+static UCS_F_ALWAYS_INLINE
+ucs_status_t ucp_do_am_zcopy_single(uct_pending_req_t *self, uint8_t am_id,
+                                    const void *hdr, size_t hdr_size,
+                                    ucp_req_complete_func_t complete)
 {
     ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
-    ucp_ep_t *ep = req->send.ep;
+    ucp_ep_t *ep       = req->send.ep;
+    size_t max_iov     = ucp_ep_config(ep)->max_am_iovcnt;
+    size_t iovcnt      = 0;
+    ucp_frag_state_t saved_state;
+    uct_iov_t iov[max_iov];
```

Comment:
i don't think it's portable, also it may be extremely large. several options to fix:
1. use malloc
2. use ucs_alloca() +have an upper limit for ucp_ep_config(ep)->max_am_iovcnt, even if the transport supports more